File: engine-dart/lib/src/core/page_definition.dart
import '../types/context.dart';
import '../types/state.dart';
import '../types/widget.dart' as w;
import 'page.dart';

/// Config for building an inline page without class inheritance.
class PageDefinitionConfig {
  final String id;
  final String title;
  final List<String> appState;
  final CachePolicy cachePolicy;
  final int cacheTtl;
  final List<StateDefinition> Function(PageContext)? state;
  final Future<dynamic> Function(PageContext)? getInfoData;
  final w.Widget Function(PageContext, dynamic) render;
  final void Function(PageContext, PageResponse)? postRender;

  const PageDefinitionConfig({
    required this.id,
    required this.title,
    this.appState = const [],
    this.cachePolicy = 'none',
    this.cacheTtl = 60,
    this.state,
    this.getInfoData,
    required this.render,
    this.postRender,
  });
}

/// Factory for creating pages from closures.
class PageDefinition {
  static Page create(PageDefinitionConfig config) => _InlinePage(config);
}

class _InlinePage extends Page {
  final PageDefinitionConfig _config;

  _InlinePage(this._config);

  @override
  String get id => _config.id;
  @override
  String get title => _config.title;
  @override
  CachePolicy get cachePolicy => _config.cachePolicy;
  @override
  int get cacheTtl => _config.cacheTtl;

  @override
  List<String> requiredAppState() => _config.appState;

  @override
  Future<dynamic> getInfoData(PageContext context) async {
    if (_config.getInfoData != null) return _config.getInfoData!(context);
    return null;
  }

  @override
  List<StateDefinition> getState(PageContext context) {
    if (_config.state != null) return _config.state!(context);
    return [];
  }

  @override
  w.Widget render(PageContext context, dynamic infoData) =>
      _config.render(context, infoData);

  @override
  void postRender(PageContext context, PageResponse response) {
    _config.postRender?.call(context, response);
  }
}

File: engine/src/core/page-definition.ts
import type { Widget } from "../types/widget";
import type { StateDefinition } from "../types/state";
import type { PageContext } from "../types/context";
import { Page, type PageResponse } from "./page";
import type { CachePolicy } from "./cache";

// ── PageDefinition Builder ─────────────────────────────────

export interface PageDefinitionConfig {
  id: string;
  title: string;
  appState?: string[];
  cachePolicy?: CachePolicy;
  cacheTtl?: number;
  state?: StateDefinition[] | ((ctx: PageContext) => StateDefinition[]);
  getInfoData?: (ctx: PageContext) => Promise<unknown> | unknown;
  render: (ctx: PageContext, infoData: unknown) => Widget;
  postRender?: (ctx: PageContext, response: PageResponse) => void;
}

export class PageDefinition {
  static create(config: PageDefinitionConfig): Page {
    return new InlinePage(config);
  }
}

class InlinePage extends Page {
  readonly id: string;
  readonly title: string;
  readonly cachePolicy: CachePolicy;
  readonly cacheTtl: number;
  private config: PageDefinitionConfig;

  constructor(config: PageDefinitionConfig) {
    super();
    this.id = config.id;
    this.title = config.title;
    this.cachePolicy = config.cachePolicy ?? "none";
    this.cacheTtl = config.cacheTtl ?? 60;
    this.config = config;
  }

  requiredAppState(): string[] {
    return this.config.appState ?? [];
  }

  async getInfoData(context: PageContext): Promise<unknown> {
    if (this.config.getInfoData) {
      return this.config.getInfoData(context);
    }
    return undefined;
  }

  getState(context: PageContext): StateDefinition[] {
    if (!this.config.state) return [];
    if (typeof this.config.state === "function") {
      return this.config.state(context);
    }
    return this.config.state;
  }

  render(context: PageContext, infoData: unknown): Widget {
    return this.config.render(context, infoData);
  }

  postRender(context: PageContext, response: PageResponse): void {
    this.config.postRender?.(context, response);
  }
}
